# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webcodecs/VideoEncoder.cpp
# Commit: e3395cf67e5c
# Full Hash: e3395cf67e5c728aaa582a2af22236156724dce5
# Author: Paul Adenot <paul@paul.cx>
# Date: 2023-12-23 09:52:03
# Regressor Bug: 1749047
# File Overlap Count: 1
# Description:
#   Bug 1749047 - Allow converting VideoEncoderConfigInternal to something a PEM can understand. r=chunmin
#   
#   It intentionally more or less maps 1:1 since my refactoring.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D195706
# ==============================================================================

diff -r bc4f18f14bd3 -r e3395cf67e5c dom/media/webcodecs/VideoEncoder.cpp
--- a/dom/media/webcodecs/VideoEncoder.cpp	Fri Dec 22 21:39:07 2023 +0000
+++ b/dom/media/webcodecs/VideoEncoder.cpp	Fri Dec 22 21:39:07 2023 +0000
@@ -154,6 +154,37 @@
   return rv;
 }
 
+EncoderConfig VideoEncoderConfigInternal::ToEncoderConfig() const {
+  MediaDataEncoder::Usage usage;
+  if (mLatencyMode == LatencyMode::Quality) {
+    usage = MediaDataEncoder::Usage::Record;
+  } else {
+    usage = MediaDataEncoder::Usage::Realtime;
+  }
+  MediaDataEncoder::HardwarePreference hwPref =
+      MediaDataEncoder::HardwarePreference::None;
+  if (mHardwareAcceleration ==
+      mozilla::dom::HardwareAcceleration::Prefer_hardware) {
+    hwPref = MediaDataEncoder::HardwarePreference::RequireHardware;
+  } else if (mHardwareAcceleration ==
+             mozilla::dom::HardwareAcceleration::Prefer_software) {
+    hwPref = MediaDataEncoder::HardwarePreference::RequireSoftware;
+  }
+  CodecType codecType = CodecType::H264;
+  Maybe<EncoderConfig::CodecSpecific> specific;
+  if (codecType == CodecType::H264) {
+    uint8_t profile, constraints, level;
+    ExtractH264CodecDetails(mCodec, profile, constraints, level);
+    specific.emplace(H264Specific(static_cast<H264_PROFILE>(profile)));
+  }
+  return EncoderConfig(
+      codecType, {mWidth, mHeight}, usage, ImageBitmapFormat::RGBA32, ImageBitmapFormat::RGBA32,
+      AssertedCast<uint8_t>(mFramerate.refOr(0.f)), 0, mBitrate.refOr(0),
+      mBitrateMode == VideoEncoderBitrateMode::Constant
+          ? MediaDataEncoder::BitrateMode::Constant
+          : MediaDataEncoder::BitrateMode::Variable,
+      hwPref, specific);
+}
 /*
  * The followings are helpers for VideoEncoder methods
  */
